static/app/components/organizations/projectPageFilter.tsx

```diff
@@ -0,0 +1,493 @@
+
+  const onClear = useCallback(() => {
+    trackAdvancedAnalyticsEvent('projectselector.clear', {
+      path: getRouteStringFromRoutes(routes),
+      organization,
+    });
+  }, [routes, organization]);
+
+  const options = useMemo<SelectOptionOrSection<number>[]>(() => {
+    const hasProjects = !!memberProjects.length || !!nonMemberProjects.length;
+    if (!hasProjects) {
+      return [];
+    }
+
+    const getProjectItem = (project: Project) => {
+      return {
+        value: parseInt(project.id, 10),
+        label: project.slug,
+        leadingItems: (
+          <ProjectBadge project={project} avatarSize={16} hideName disableLink />
+        ),
+        trailingItems: ({isFocused}) => (
+          <Fragment>
+            <TrailingButton
+              borderless
+              size="zero"
+              icon={<IconOpen />}
+              aria-label={t('Project Details')}
+              to={`/organizations/${organization.slug}/projects/${project.slug}/?project=${project.id}`}
+              visible={isFocused}
+            />
+            <TrailingButton
+              borderless
+              size="zero"
+              icon={<IconSettings />}
+              aria-label={t('Project Settings')}
+              to={`/settings/${organization.slug}/projects/${project.slug}/`}
+              visible={isFocused}
+            />
+            <StyledBookmarkStar
+              project={project}
+              organization={organization}
+              visible={isFocused}
+              onToggle={(isBookmarked: boolean) => {
+                analytics('projectselector.bookmark_toggle', {
+                  org_id: parseInt(organization.id, 10),
+                  bookmarked: isBookmarked,
+                });
+              }}
+            />
+          </Fragment>
+        ),
+      };
+    };
+
+    const lastSelected = mapURLValueToNormalValue(pageFilterValue);
+    const listSort = (project: Project) => [
+      !lastSelected.includes(parseInt(project.id, 10)),
+      !project.isBookmarked,
+      project.slug,
+    ];
+
+    return nonMemberProjects.length > 0
+      ? [
+          {
+            key: 'my-projects',
+            label: t('My Projects'),
+            options: sortBy(memberProjects, listSort).map(getProjectItem),
+            showToggleAllButton: allowMultiple,
+          },
+          {
+            key: 'no-membership-header',
+            label:
+              memberProjects.length > 0 ? t('Other') : t("Projects I Don't Belong To"),
+            options: sortBy(nonMemberProjects, listSort).map(getProjectItem),
+            showToggleAllButton: allowMultiple && memberProjects.length > 0,
+          },
+        ]
+      : sortBy(memberProjects, listSort).map(getProjectItem);
+  }, [
+    organization,
+    allowMultiple,
+    memberProjects,
+    nonMemberProjects,
+    mapURLValueToNormalValue,
+    pageFilterValue,
+  ]);
+
+  const trigger = useCallback<NonNullable<MultipleSelectProps<number>['trigger']>>(
```

Comment:
nit: these `useMemo`s and `useCallback`s are petty long, do you think we could extract them into their own hooks/components instead of inlining?

Reply:
Ah good point! Did so in https://github.com/getsentry/sentry/pull/47195/commits/afe21462b8d2ac72fef0fc08767bc2b45c709f09